Fixed #35671 -- Clarified string-based fields behavior when null=False.

cliff688 opened the pull request:
# Trac ticket number
<!-- Replace 35671 with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-35671

# Branch description
This branch makes improvements to the documentation of ``Field.null`` and ``Field.default``
clarifying the behaviour of these model options on string-based fields, particularly
what happens when ``null=False`` and ``default`` is left unspecified for these fields.

The current docs do not mention these cases. The case is significant because
unlike other types of fields, ``null=False`` does not have the effect of disallowing empty
values for string-based fields. Empty strings can still be saved. This happens when
a default is unspecified and Django uses an empty string. This is also not clearly
documented.

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

sarahboyce commented on docs/ref/models/fields.txt:
```diff
@@ -51,6 +51,12 @@ for "no data;" the Django convention is to use the empty string, not
 and ``blank=True`` set. In this situation, ``null=True`` is required to avoid
 unique constraint violations when saving multiple objects with blank values.
 
+Additionally, be cautious of depending on :attr:`~Field.null` to enforce a
+"not null" constraint at the database level for string-based fields, as
+databases may accept the empty string as a value despite :attr:`~Field.null`
+being ``False``. To ensure that a string-based field is not empty, you should
+set :attr:`default=None <Field.default>` and :attr:`null=False <Field.null>`.
```
I'm not sure about this because this does enforce a not null constraint (it won't save null values)

I wonder if this is more about `blank=False` and saying that `blank=False` does not mean you won't be able to save blank values. You need to call `full_clean()` to run the validation

cliff688 replied:
Hello Sarah! Thank you for the review.
Perhaps this may be about ``blank=False`` too. But I hadn't thought about that since I experienced the issue when I created model instances in my views or in the interactive shell using ``create()`` or ``get_or_create()``, which is why I thought it was about ``null=False`` behavior on string-based fields when no ``default`` is specified. Specifically, how ``null=False`` will not disallow saving when no value is provided for sting-based field, which happens because database accepts empty strings as value.
Please, let me know what you think.

sarahboyce replied:
Alright, let's keep the note in the null docs :+1:

sarahboyce commented on docs/ref/models/fields.txt:
```diff
@@ -479,7 +485,9 @@ should be the value of the field they reference (``pk`` unless
+(see :attr:`~Field.null`).
 
 The default value can also be set at the database level with
 :attr:`Field.db_default`.
```
If we put anything here I think maybe we should have something like
```
See :meth:`Field.get_default` for how the default value is determined when
``default`` is not set.
```
and further down under the `descriptor_class` docs
```

    .. method:: get_default()

        Returns the default value of the field. This is :attr:`~Field.default`
        or :attr:`~Field.db_default`, with :attr:`~Field.default` taking
        precedence over :attr:`Field.db_default` if both are set. If neither are
        set, then returns an empty string if the field allows it, otherwise
        returns ``None``.
```

But I'm not sure if a comment about `blank` might be enough

cliff688 replied:
Yes I agree. That's certainly better than my addition in ``Field.default``. I wasn't able to find the docs for ``Field.get_default`` previously.

cliff688 replied:
I'm not sure who should add the documentation for ``Field.get_default()``. Would you like to handle it, or should I include it in the Pull Request?

sarahboyce replied:
I've decided to remove `Field.get_default()` from the pull request for now as I am not 100% sure whether the clarity we gain is worth adding this to the public API :thinking: 
But moving forward with the other changes :+1:

sarahboyce reviewed (commented):
Thank you for starting this @cliff688 :star:
I think this is quite fiddly, so I have added a couple of comments, let me know what you think
I think it might take a few iterations before we come up with something we're happy with

timgraham commented on docs/ref/models/fields.txt:
```diff
@@ -51,6 +51,12 @@ for "no data;" the Django convention is to use the empty string, not
 
+Additionally, be cautious of depending on :attr:`~Field.null` to enforce a
+"not null" constraint at the database level for string-based fields, as
+databases may accept the empty string as a value despite :attr:`~Field.null`
```
I think "may accept" needs more precision. I guess it relates to the note below about Oracle storing null if you pass it an empty string? It would be interesting to know if any other databases (outside of Django's built-in backends) do that.

cliff688 replied:
Hi @timgraham,

Thank you for the review! I understand your concern about the precision of the "may accept".

Should I clarify that among Django's built-in database backends PostgreSQL, MySQL, and SQLite, consider empty strings as valid string values? (The case for Oracle is already documented in the note on ``Field.null``.)

sarahboyce replied:
There are more hints about the empty string behavior than I first thought

We could change this paragraph to be:
```
That ``null=False`` has no impact on whether empty strings are allowed for
string-based fields. If you wish to prevent empty strings being saved to the
database, you need to call ``full_clean()`` before saving a model object when
:attr:`blank=False <Field.blank>`. You may also want to set :attr:`default=None
<Field.default>`, as the Django convention is to use empty strings by default.
```
The "as the Django convention is to use empty strings by default" is also said above but I think it's useful to reiterate it

cliff688 replied:
Hello @sarahboyce! Sorry, I saw this after I had made and pushed my most recent commits. I am considering replacing the admonition I just included in favour of this. What do you think? Should I replace the advice with CheckConstraint in favour of this? Myself I prefer the ``CheckConstraint`` over calling full-clean() on every save.

sarahboyce replied:
I read the admonition. I'm not sure we can say what "the preferred method" is as there are different ways to approach this
Maybe just adding ```Empty strings can still be saved to the database when ``null=False``.``` at the end of the paragraph about string based fields is enough?

cliff688 replied:
Okay I'll do that. Thanks.

cliff688 replied:
On second thought, I think we still have to add some pointers for how one might ensure that empty strings are not saved in database so readers know where to look for solutions. Here is what I propose:
After the second paragraph:
```reST
Empty strings can still be saved to the database when ``null=False``. If you
wish to disallow them, one method is to use a :class:`CheckConstraint`.
Alternatively, you could add extra validation logic to the model's
:meth:`~Model.clean()` method.
```
Let me know what you think.

sarahboyce replied:
Been thinking about this some more and I think maybe it's enough to move the comment of `the Django convention is to use the empty string, not NULL` further up so

```
Avoid using :attr:`~Field.null` on string-based fields such as
:class:`CharField` and :class:`TextField`. The Django convention is to use an
empty string, not ``NULL`` as the "no data" state for string-based fields. If a
string-based field has ``null=True``, that means it has two possible values for
"no data": ``NULL``, and the empty string. In most cases, it's redundant to have
two possible values for "no data". One exception is when a :class:`CharField`
has both ``unique=True`` and ``blank=True`` set. In this situation,
``null=True`` is required to avoid unique constraint violations when saving
multiple objects with blank values.
```

And update
```diff
 .. attribute:: Field.null
 
-If ``True``, Django will store empty values as ``NULL`` in the database. Default
-is ``False``.
+If ``True``, Django can store ``NULL`` values in the database. Default is
+``False``.
```
as that might be slightly more acurate

cliff688 replied:
I think the paragraph with "the Django convention" you propose is indeed more effective than the previous one. But I'm not so sure about
```diff
+If ``True``, Django can store ``NULL`` values in the database. Default is
+``False``.
```
Even for string-based fields (as far as I have been able to check) if ``null=True`` Django will in fact save ``NULL`` in the database for "no data". So maybe we ought to let that part alone?
The inconsistencies only happens when ``null=False``, which is where our problem lies

sarahboyce replied:
Yes sorry :+1: makes sense

cliff688 replied:
How about modifying the passage on string fields to:

```diff
Avoid using :attr:`~Field.null` on string-based fields such as
:class:`CharField` and :class:`TextField`. The Django convention is to use an
empty string, not ``NULL`` as the "no data" state for string-based fields.
+ If a string based-field has ``null=False`` empty strings can still be saved.
If a string-based field has ``null=True``, that means it has two possible values
for "no data": ``NULL``, and the empty string. In most cases, it's redundant to
have two possible values for "no data". One exception is when a
:class:`CharField` has both ``unique=True`` and ``blank=True`` set. In this
situation, ``null=True`` is required to avoid unique constraint violations when
saving multiple objects with blank values.
```

cliff688 commented:
I have updated the docs with changes suggested by @sarahboyce and also addressed the concerns about the clarity of my documentation on ``Field.null``. Please let me know what you think.

cliff688 commented:
I have updated the docs with the agreed-upon changes.
Please let me know what you think.

cliff688 commented:
Hello!
Is it okay to rebase commits?

timgraham commented:
Yes, rebase is preferred.

sarahboyce reviewed (approved):
Thank you for your work on this @cliff688 :star:

timgraham commented:
If I'm reading this correctly, the only change is adding:  

`If a string based-field has ``null=False``, empty strings can still be saved for "no data".`

I think it's surprising to have to state that explicitly, but I guess it's fine. It doesn't seem to address the ticket's summary: "Document how to disallow blank values in Char/TextField".

By the way, there's a typo in the dash position of "string based-field". (should be: "string-based")

sarahboyce commented:
> If I'm reading this correctly, the only change is adding:
> 
> `` If a string based-field has ``null=False``, empty strings can still be saved for "no data". ``

Correct

> I think it's surprising to have to state that explicitly, but I guess it's fine. It doesn't seem to address the ticket's summary: "Document how to disallow blank values in Char/TextField".

You're right and so I have updated this to "Clarify behavior for Field.null=False on string-based fields". 
The original ticket summary was:
"Field.null=False on string-based fields behaves inconsistently with the typical usage of Field.null."

I had updated the summary to  "Document how to disallow blank values in Char/TextField" because that's where I thought the confusion was, but in the review I think the core was `null=False` doesn't mean `"no data allowed"=False`

> By the way, there's a typo in the dash position of "string based-field". (should be: "string-based")

Thank you! Updated :+1:
